File: ros2_packages/lapgym_ros2_bridge/lapgym_ros2_bridge/hold_policy_server.py
"""
Phase 4B: HoldPolicyServer -- holds instrument at current position.

After RetractPolicyServer completes, this server keeps the instrument
stationary by publishing zero-delta actions. The surgeon can inspect
the retracted tissue. Runs until cancelled by the Behaviour Tree.

Author: Subhash Arockiadoss
"""

# -- Gymnasium shim MUST be first ---------------------------------------------
import sys
import gymnasium
sys.modules['gym'] = gymnasium
sys.modules['gym.spaces'] = gymnasium.spaces
# -----------------------------------------------------------------------------

import os
import threading
import time
import numpy as np
import rclpy
from std_msgs.msg import Bool
from rclpy.node import Node
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup

from lapgym_interfaces.action import Retract

DEFAULT_MAX_STEPS = 500   # hold for up to 500 steps (~55 seconds)
HOLD_STEP_DELAY = 0.1     # 10 Hz hold rate -- no need for 50Hz


class HoldPolicyServer(Node):
    """Holds instrument at current position by publishing zero actions."""

    def __init__(self):
        super().__init__('hold_policy_server')

        self._env = None
        self._load_env()

        self._action_server = ActionServer(
            self,
            Retract,
            'hold_policy',
            execute_callback=self._execute_cb,
            goal_callback=self._goal_cb,
            cancel_callback=self._cancel_cb,
            callback_group=ReentrantCallbackGroup(),
        )

        # # Surgeon stop -- dedicated background node + thread
        # # env.step() blocks the main executor so we need a separate thread
        self._surgeon_stopped = False
        self._stop_event = threading.Event()
        self._stop_event.set()
        self._emergency = False

        self._stop_context = rclpy.Context()
        self._stop_context.init()

        self._stop_node = rclpy.create_node(
            '_surgeon_stop_hold',
            context=self._stop_context,
            enable_rosout=False
        )
        self._stop_node.create_subscription(
            Bool, '/surgeon_stop', self._surgeon_cb, 10)
        self._stop_node.create_subscription(
            Bool, '/emergency_stop', self._emergency_cb, 10)

        self._stop_executor = rclpy.executors.SingleThreadedExecutor(
            context=self._stop_context)
        self._stop_executor.add_node(self._stop_node)

        self._stop_thread = threading.Thread(
            target=self._spin_stop_node, daemon=True)
        self._stop_thread.start()


        self.get_logger().info('HoldPolicyServer ready')

    def _spin_stop_node(self):
        while self._stop_context.ok():
            try:
                self._stop_executor.spin_once(timeout_sec=0.01)
            except Exception:
                pass

    def _load_env(self):
        try:
            from sofa_env.scenes.tissue_retraction.tissue_retraction_env import RenderMode
            from envs.tissue_retraction_v2 import TissueRetractionV2
            self._env = TissueRetractionV2(
                env_kwargs={'render_mode': RenderMode.HEADLESS})
            obs, _ = self._env.reset()
            self._obs = obs
            self.get_logger().info('Environment ready')
        except Exception as e:
            self.get_logger().error(f'Failed to load env: {e}')
            self._env = None

    def _goal_cb(self, goal_request):
        if self._env is None:
            return GoalResponse.REJECT
        return GoalResponse.ACCEPT

    def _cancel_cb(self, goal_handle):
        self.get_logger().info('Hold cancel received -- accepting')
        return CancelResponse.ACCEPT
    
    def _emergency_cb(self, msg: Bool):
        if msg.data and not self._emergency:
            self._emergency = True
            self.get_logger().error(
                'Hold server: EMERGENCY STOP received -- halting')
            
    # add callback
    def _surgeon_cb(self, msg: Bool):
        if msg.data:
            self._surgeon_stopped = True
            self._stop_event.clear()
            self.get_logger().warn('Hold: SURGEON STOP received')
        else:
            self._surgeon_stopped = False
            self._stop_event.set()
            self.get_logger().info('Hold: SURGEON RESUME received')
    async def _execute_cb(self, goal_handle):
        self.get_logger().info('Hold policy active -- holding position')

        goal = goal_handle.request
        max_steps = int(goal.max_steps) if goal.max_steps > 0 else DEFAULT_MAX_STEPS

        # Zero action -- hold current position
        zero_action = np.zeros(3, dtype=np.float32)
        feedback_msg = Retract.Feedback()
        step = 0

        while step < max_steps:

            # -- Preemption check FIRST ---------------------------------------
            if goal_handle.is_cancel_requested:
                goal_handle.canceled()
                self.get_logger().info(f'Hold preempted at step {step}')
                result = Retract.Result()
                result.success = True   # hold is always "successful"
                result.steps_taken = step
                result.final_distance = 0.0
                result.termination = 'preempted'
                return result
            
            # -- Emergency stop check -----------------------------------------
            if self._emergency:
                self.get_logger().error(
                    f'Hold emergency stop at step {step}')
                try:
                    goal_handle.canceled()
                except Exception:
                    goal_handle.abort()
                result = Retract.Result()
                result.success = False
                result.steps_taken = step
                result.final_distance = 0.0
                result.termination = 'emergency_stop'
                return result
            
            # -- Surgeon stop: freeze until resumed ---------------------------
            while self._surgeon_stopped and not self._emergency:
                if goal_handle.is_cancel_requested:
                    break
                self._stop_event.wait(timeout=0.05)
            
            # -- Publish zero action to hold position -------------------------
            obs, reward, terminated, truncated, info = self._env.step(zero_action)
            step += 1

            # --Immediate Surgeon stop: freeze until resumed ---------------------------
            while self._surgeon_stopped and not self._emergency:
                if goal_handle.is_cancel_requested:
                    break
                self._stop_event.wait(timeout=0.05)

            # -- Feedback every step (for accurate console display)
            
            in_collision = bool(info.get('in_collision', False))
            collision_cost = abs(float(info.get('collision_cost', 0.0) or 0.0))
            feedback_msg.distance_to_goal = 0.0
            feedback_msg.distance_mm = 0.0
            feedback_msg.step = step
            feedback_msg.in_collision = in_collision
            feedback_msg.collision_cost = collision_cost
            goal_handle.publish_feedback(feedback_msg)
    
            if step % 5 == 0:
                self.get_logger().info(
                    f'Holding step {step:3d} | '
                    f'{"COL" if in_collision else "SAFE"}')

            time.sleep(HOLD_STEP_DELAY)

        goal_handle.succeed()
        result = Retract.Result()
        result.success = True
        result.steps_taken = step
        result.final_distance = 0.0
        result.termination = 'timeout'
        self.get_logger().info('Hold timeout reached')
        return result

    def destroy_node(self):
        self._stop_executor.shutdown()
        self._stop_node.destroy_node()
        self._stop_context.shutdown()
        if self._env is not None:
            try:
                self._env.close()
            except Exception:
                pass
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = HoldPolicyServer()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()
        os._exit(0)

File: ros2_packages/lapgym_ros2_bridge_cpp/src/hold_policy_server.cpp
/**
 * Phase 4G: HoldPolicyServer (C++)
 *
 * Holds instrument at current position by sending zero-delta actions.
 * Simplest C++ action server — no ML, no PPO, pure control logic.
 *
 * Demonstrates same safety patterns as approach_policy_server.cpp:
 *   - std::atomic<bool> for surgeon stop and emergency flags
 *   - Separate callback group for stop listener
 *   - MultiThreadedExecutor for concurrent execution
 *   - Dual freeze loop (before + after sofa step)
 *
 * Author: Subhash Arockiadoss
 */

#include <chrono>
#include <memory>
#include <thread>
#include <atomic>
#include <array>
#include <string>
#include <functional>
#include <future>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "std_msgs/msg/bool.hpp"

#include "lapgym_interfaces/action/retract.hpp"
#include "lapgym_interfaces/srv/sofa_step.hpp"

static constexpr int DEFAULT_MAX_STEPS = 500;

using Retract           = lapgym_interfaces::action::Retract;
using SofaStep          = lapgym_interfaces::srv::SofaStep;
using GoalHandleRetract = rclcpp_action::ServerGoalHandle<Retract>;


class HoldPolicyServerCpp : public rclcpp::Node
{
public:
    HoldPolicyServerCpp()
    : Node("hold_policy_server_cpp"),
      surgeon_stopped_(false),
      emergency_(false)
    {
        // -- Action server -------------------------------------------------
        action_server_ = rclcpp_action::create_server<Retract>(
            this,
            "hold_policy_cpp",
            std::bind(&HoldPolicyServerCpp::handleGoal,     this,
                      std::placeholders::_1, std::placeholders::_2),
            std::bind(&HoldPolicyServerCpp::handleCancel,   this,
                      std::placeholders::_1),
            std::bind(&HoldPolicyServerCpp::handleAccepted, this,
                      std::placeholders::_1)
        );

        // -- SOFA step service client ---------------------------------------
        client_callback_group_ = create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);
        sofa_client_ = create_client<SofaStep>(
            "/sofa_step",
            rmw_qos_profile_services_default,
            client_callback_group_);

        RCLCPP_INFO(get_logger(), "Waiting for /sofa_step service...");
        while (!sofa_client_->wait_for_service(std::chrono::seconds(1))) {
            if (!rclcpp::ok()) return;
        }
        RCLCPP_INFO(get_logger(), "/sofa_step service ready");

        // -- Stop listener on background thread ----------------------------
        stop_callback_group_ = create_callback_group(
            rclcpp::CallbackGroupType::MutuallyExclusive);

        rclcpp::SubscriptionOptions stop_opts;
        stop_opts.callback_group = stop_callback_group_;

        surgeon_stop_sub_ = create_subscription<std_msgs::msg::Bool>(
            "/surgeon_stop", 10,
            std::bind(&HoldPolicyServerCpp::surgeonStopCb, this,
                      std::placeholders::_1),
            stop_opts);

        emergency_sub_ = create_subscription<std_msgs::msg::Bool>(
            "/emergency_stop", 10,
            std::bind(&HoldPolicyServerCpp::emergencyCb, this,
                      std::placeholders::_1),
            stop_opts);

        stop_executor_ =
            std::make_shared<rclcpp::executors::SingleThreadedExecutor>();
        stop_executor_->add_callback_group(
            stop_callback_group_, get_node_base_interface());
        stop_thread_ = std::thread([this]() {
            stop_executor_->spin();
        });

        RCLCPP_INFO(get_logger(),
            "HoldPolicyServerCpp ready on /hold_policy_cpp");
    }

    ~HoldPolicyServerCpp()
    {
        if (stop_executor_) stop_executor_->cancel();
        if (stop_thread_.joinable()) stop_thread_.join();
    }

private:
    // -----------------------------------------------------------------------
    // Action callbacks
    // -----------------------------------------------------------------------

    rclcpp_action::GoalResponse handleGoal(
        const rclcpp_action::GoalUUID & /*uuid*/,
        std::shared_ptr<const Retract::Goal> /*goal*/)
    {
        RCLCPP_INFO(get_logger(), "Hold goal received — accepting");
        return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
    }

    rclcpp_action::CancelResponse handleCancel(
        const std::shared_ptr<GoalHandleRetract> /*goal_handle*/)
    {
        RCLCPP_INFO(get_logger(), "Hold cancel — accepting");
        return rclcpp_action::CancelResponse::ACCEPT;
    }

    void handleAccepted(const std::shared_ptr<GoalHandleRetract> goal_handle)
    {
        std::thread([this, goal_handle]() {
            execute(goal_handle);
        }).detach();
    }

    // -----------------------------------------------------------------------
    // Main execution loop — zero actions, hold position
    // -----------------------------------------------------------------------

    void execute(const std::shared_ptr<GoalHandleRetract> goal_handle)
    {
        RCLCPP_INFO(get_logger(), "Hold policy active — holding position");

        auto goal     = goal_handle->get_goal();
        int max_steps = (goal->max_steps > 0)
                        ? static_cast<int>(goal->max_steps)
                        : DEFAULT_MAX_STEPS;

        auto feedback = std::make_shared<Retract::Feedback>();
        int  step     = 0;

        while (step < max_steps) {

            // -- Preemption check ------------------------------------------
            if (goal_handle->is_canceling()) {
                auto result            = std::make_shared<Retract::Result>();
                result->success        = true;   // hold always succeeds
                result->steps_taken    = step;
                result->final_distance = 0.0f;
                result->termination    = "preempted";
                goal_handle->canceled(result);
                RCLCPP_INFO(get_logger(), "Hold preempted at step %d", step);
                return;
            }

            // -- Emergency stop --------------------------------------------
            if (emergency_.load()) {
                auto result            = std::make_shared<Retract::Result>();
                result->success        = false;
                result->steps_taken    = step;
                result->final_distance = 0.0f;
                result->termination    = "emergency_stop";
                try { goal_handle->abort(result); } catch (...) {}
                RCLCPP_ERROR(get_logger(),
                    "Hold EMERGENCY STOP at step %d", step);
                return;
            }

            // -- Surgeon stop freeze loop ----------------------------------
            while (surgeon_stopped_.load() && !emergency_.load()) {
                if (goal_handle->is_canceling()) break;
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }

            // -- Send zero action — hold position --------------------------
            // Zero delta means instrument does not move
            // SOFA physics still advances but position unchanged
            sofaStep({0.0f, 0.0f, 0.0f});
            step++;

            // -- Post-step freeze loop -------------------------------------
            while (surgeon_stopped_.load() && !emergency_.load()) {
                if (goal_handle->is_canceling()) break;
                std::this_thread::sleep_for(std::chrono::milliseconds(50));
            }

            // -- Feedback every step ---------------------------------------
            feedback->distance_to_goal = 0.0f;
            feedback->distance_mm      = 0.0f;
            feedback->step             = step;
            feedback->in_collision     = false;
            feedback->collision_cost   = 0.0f;
            goal_handle->publish_feedback(feedback);

            if (step % 5 == 0) {
                RCLCPP_INFO(get_logger(),
                    "Holding (C++) step %3d", step);
            }

            // Hold at 10Hz — no need for full speed
            std::this_thread::sleep_for(std::chrono::milliseconds(100));
        }

        // -- Timeout — normal completion -----------------------------------
        auto result            = std::make_shared<Retract::Result>();
        result->success        = true;
        result->steps_taken    = step;
        result->final_distance = 0.0f;
        result->termination    = "timeout";
        goal_handle->succeed(result);
        RCLCPP_INFO(get_logger(), "Hold timeout reached at step %d", step);
    }

    // -----------------------------------------------------------------------
    // SOFA service helper
    // -----------------------------------------------------------------------

    void sofaStep(std::array<float, 3> action)
    {
        auto request    = std::make_shared<SofaStep::Request>();
        request->action = action;
        request->reset  = false;

        auto future = sofa_client_->async_send_request(request);
        future.wait_for(std::chrono::seconds(5));
        // result not needed for hold — we only care that step happened
    }

    // -----------------------------------------------------------------------
    // Stop callbacks
    // -----------------------------------------------------------------------

    void surgeonStopCb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data) {
            surgeon_stopped_.store(true);
            RCLCPP_WARN(get_logger(), "Hold (C++): SURGEON STOP received");
        } else {
            surgeon_stopped_.store(false);
            RCLCPP_INFO(get_logger(), "Hold (C++): SURGEON RESUME received");
        }
    }

    void emergencyCb(const std_msgs::msg::Bool::SharedPtr msg)
    {
        if (msg->data && !emergency_.load()) {
            emergency_.store(true);
            RCLCPP_ERROR(get_logger(),
                "Hold (C++): EMERGENCY STOP received");
        }
    }

    // -----------------------------------------------------------------------
    // Member variables
    // -----------------------------------------------------------------------

    rclcpp_action::Server<Retract>::SharedPtr                    action_server_;
    rclcpp::Client<SofaStep>::SharedPtr                          sofa_client_;
    rclcpp::CallbackGroup::SharedPtr                             client_callback_group_;
    std::atomic<bool>                                            surgeon_stopped_;
    std::atomic<bool>                                            emergency_;
    rclcpp::CallbackGroup::SharedPtr                             stop_callback_group_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         surgeon_stop_sub_;
    rclcpp::Subscription<std_msgs::msg::Bool>::SharedPtr         emergency_sub_;
    std::shared_ptr<rclcpp::executors::SingleThreadedExecutor>   stop_executor_;
    std::thread                                                  stop_thread_;
};


int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HoldPolicyServerCpp>();

    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();

    rclcpp::shutdown();
    return 0;
}